### PR TITLE
requireTemplateStrings: check if left/right is a string

### LINKED ISSUE #1416: requireTemplateStrings: add more checks
Right now we check for `Identifier` and `BinaryExpression`

There might be more based on these?

```js
this.toIcon(suit) + ' Card-suit'; // node.left.type === 'CallExpression'
card.value + ' is a point card'; // node.left.type === 'MemberExpression'
```

Ref original: https://github.com/jscs-dev/node-jscs/issues/1330

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/18951009-requiretemplatestrings-add-more-checks?utm_campaign=plugin&utm_content=tracker%2F281640&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F281640&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by hzoo**
Fixes #1416

Did a little refactor for if we need to add more types?

### CODE REVIEW

File: lib/rules/require-template-strings.js [hunk 1]
```diff
@@ -67,23 +80,11 @@ module.exports.prototype = {
...
-
-            // (a + b) + "a"
-            if (node.left.type === 'BinaryExpression') {
+            if (nodeTypes[node.left.type]) {
                 add(node);
             }
 
-            // "a" + (a + b)
-            if (node.right.type === 'BinaryExpression') {
+            if (nodeTypes[node.right.type]) {
```
Comment (mrjoelkemp) [inline lib/rules/require-template-strings.js]: Combine conditionals?

File: test/specs/rules/require-template-strings.js [hunk 2]
```diff
@@ -42,6 +42,20 @@ describe('rules/require-template-strings', function() {
...
+    it('should report for the use of string concatenation with a CallExpression',
+        function() {
+            assert(checker.checkString('a() + "a"').getErrorCount() === 1);
+            assert(checker.checkString('"a" + a()').getErrorCount() === 1);
+        }
+    );
+
+    it('should report for the use of string concatenation with a MemberExpression',
+        function() {
+            assert(checker.checkString('a.b + "a"').getErrorCount() === 1);
```
Comment (mrjoelkemp) [inline test/specs/rules/require-template-strings.js]: What about a template string concatenated with a regular string. Is that possible?

Comment (mrjoelkemp) [inline test/specs/rules/require-template-strings.js]: Also, can you add tests that have more than just two operands being concatenated?

Comment (hzoo) [inline test/specs/rules/require-template-strings.js]: Yeah don't know why anyone would but we can report for that too.

Comment (mrjoelkemp) [inline test/specs/rules/require-template-strings.js]: What about for tagged template strings being concatenated with another string or template string. I guess I'm questioning the ast type for a tagged template string.

Comment (hzoo) [inline test/specs/rules/require-template-strings.js]: Yeah its different so have to do both.

Comment (hzoo) [inline test/specs/rules/require-template-strings.js]: > Also, can you add tests that have more than just two operands being concatenated?

I guess that would be what @markelog added before

```js
it('should report the use of string concatenation with right handed binary expression',
        function() {
            assert(checker.checkString('"test" + (a + b)').getErrorCount() === 1);
        }
    );

    it('should report the use of string concatenation with left handed binary expression',
        function() {
            assert(checker.checkString('(a + b) + "test"').getErrorCount() === 1);
        }
    );
```

Comment (mrjoelkemp) [inline test/specs/rules/require-template-strings.js]: Right. Didn't catch those tests. :+1:

**Comment by mikesherov** [comment]
Can we just check for string literal + non string literal? Why have a list of types?

**Comment by hzoo** [comment]
I guess there's no reason not to - at first I was specifically checking for string and identifier. It would certainly be simpler!

**Comment by hzoo** [comment]
Do we want to make an exception for a multi-line concatenation if we do string + string?
```js
let reallyLong = 'asdfadfdsfasdfadfadfasdf' +
      'adsfasdf';
```

**Comment by markelog** [comment]
> Can we just check for string literal + non string literal? Why have a list of types?

Gush, cool.

> Do we want to make an exception for a multi-line concatenation if we do string + string?

I would say no, we always can wait for the community response on that.

**Comment by hzoo** [comment]
Not sure why the coverage on this decreased (I removed lines but it's still 100%)

File: lib/rules/require-template-strings.js [hunk 3]
```diff
@@ -62,28 +64,8 @@ module.exports.prototype = {
...
-            // (a + b) + "a"
-            if (node.left.type === 'BinaryExpression') {
-                add(node);
-            }
-
-            // "a" + (a + b)
-            if (node.right.type === 'BinaryExpression') {
+            // Only one of the operands should be a string, otherwise this is not a concatenation
+            if (typeof node.left.value === 'string' || typeof node.right.value === 'string') {
                 add(node);
```
Comment (markelog) [inline lib/rules/require-template-strings.js]: Now we can inline this

Comment (hzoo) [inline lib/rules/require-template-strings.js]: Will that fix the coverage issue? We still need the curly braces since we have that rule on.

**Comment by hzoo** [comment]
Yeah not sure the problem with coverage but should be good now.

**Comment by mrjoelkemp** [comment]
LGTM. Restarting the travis build to see if coverage passes.

**Comment by mrjoelkemp** [comment]
Seeing a travis failure now: 

image

We could try running unit-coverage locally via one of our package.json scripts to see if the error can be reproduced.

**Comment by hzoo** [comment]
Is there anything I need to do for this?

**Comment by markelog** [comment]
Not sure what the problem with coveralls here.

/cc @mdevils

**Comment by mdevils** [comment]
Looks like it's not coveralls. I will investigate with issue.

**Comment by mdevils** [comment]
Actually is looks like coveralls cannot find some commit.

**Comment by mdevils** [comment]
Looks like it could not track rebase or so. Can you try to recreate PR?

**Comment by zxqfox** [comment]
Did you tried to make a new commit? Like `git ci --amend; git push -f`

**Comment by hzoo** [comment]
I'l try a new commit, and if that doesn't work a new PR

**Comment by markelog** [comment]
Funny, now we have a test failure :-)

**Comment by hzoo** [comment]
Yep I must of did all that wrong..

**Comment by hzoo** [comment]
Oh the test that failed is the 2 string literal one - do we want to report for this?

`checker.checkString('"a" + "a"')`

**Comment by hzoo** [comment]
Wait never mind that was a duplicated test that was wrong anyway.

**Comment by hzoo** [comment]
Ok back to the old coveralls issue ` Coverage decreased (-0.0%) to 98.65%`

**Comment by markelog** [comment]
Finally :-)

**Comment by hzoo** [comment]
:tada: yay

**Pull request closed without merging**